Upgrade panic when comparing values with overrides

This is caused by some compare paths are deeper than overrides, but not supported by GetFromTreePath()

The error and CLI command:


```
taohe@taohe:~/istio.io/operator$ istioctl140b2 x upgrade --wait --skip-confirmation --force --dry-run --logtostderr -f /usr/local/google/home/taohe/default.yaml

2019-11-06T04:17:50.142989Z	info	Client - istioctl version: 1.4.0

2019-11-06T04:17:50.170244Z	info	ReadProfileYAML for profile name: [Empty]
2019-11-06T04:17:50.170271Z	info	Loading values from compiled in VFS at path profiles/default.yaml
2019-11-06T04:17:50.320401Z	info	ReadProfileYAML for profile name: [Empty]
2019-11-06T04:17:50.320431Z	info	Loading values from compiled in VFS at path profiles/default.yaml
2019-11-06T04:17:50.397257Z	info	Warning. The target version 1.4-dev does not equal to the binary version 1.4.0
2019-11-06T04:17:50.397288Z	info	Upgrade - target version: 1.4-dev

2019-11-06T04:17:50.883310Z	info	Control Plane - citadel pod - istio-citadel-6c4ccc79bc-ghpbb - version: 1.4.0-beta.2
2019-11-06T04:17:50.883344Z	info	Control Plane - galley pod - istio-galley-5fb696ff98-m5dxv - version: 1.4.0-beta.2
2019-11-06T04:17:50.883354Z	info	Control Plane - ingressgateway pod - istio-ingressgateway-5477d4d87f-j7z8f - version: 1.4.0-beta.2
2019-11-06T04:17:50.883362Z	info	Control Plane - pilot pod - istio-pilot-568d4bbf9c-kwb4s - version: 1.4.0-beta.2
2019-11-06T04:17:50.883370Z	info	Control Plane - policy pod - istio-policy-79cc59bd9-64fhx - version: 1.4.0-beta.2
2019-11-06T04:17:50.883378Z	info	Control Plane - sidecar-injector pod - istio-sidecar-injector-5f65dc5999-9fj7r - version: 1.4.0-beta.2
2019-11-06T04:17:50.883394Z	info	Control Plane - telemetry pod - istio-telemetry-bd95dc7b8-5jwts - version: 1.4.0-beta.2
2019-11-06T04:17:50.883402Z	info	
2019-11-06T04:17:51.317075Z	info	Upgrade version check passed: 1.4.0-beta.2 -> 1.4-dev.

panic: interface conversion: interface {} is string, not map[string]interface {}

goroutine 1 [running]:
istio.io/operator/pkg/name.GetFromTreePath(0xc000e58900, 0xc000eec8a0, 0x2, 0x2, 0xc000e5ca98, 0x1, 0xcd0000c0009ced36, 0xc0006e0680, 0x3)
	istio.io/operator@v0.0.0-20191105031444-4d7f1ed76494/pkg/name/name.go:235 +0x4bd
istio.io/operator/pkg/name.GetFromTreePath(0xc000e58720, 0xc000eec890, 0x3, 0x3, 0xc000e5d5e8, 0x1, 0xc0009ced27, 0xc000eec880, 0xc000df0a70)
	istio.io/operator@v0.0.0-20191105031444-4d7f1ed76494/pkg/name/name.go:232 +0x361
istio.io/operator/pkg/name.GetFromTreePath(0xc000ab74a0, 0xc000eec880, 0x4, 0x4, 0x4, 0x4, 0x217eac0, 0xc000c1ee20, 0xc000e747c0)
	istio.io/operator@v0.0.0-20191105031444-4d7f1ed76494/pkg/name/name.go:232 +0x361
istio.io/operator/pkg/compare.genYamlIgnoreOpt.func1(0xc000df0a00, 0x8, 0x10, 0xc000e74800)
	istio.io/operator@v0.0.0-20191105031444-4d7f1ed76494/pkg/compare/compare.go:184 +0x7b
github.com/google/go-cmp/cmp.pathFilter.filter(0xc000e4d220, 0x2a1d860, 0x410e7e0, 0xc00067ad80, 0x2ae8f00, 0x21ff240, 0x21ff240, 0xc000c3fc80, 0x194, 0x0, ...)
	github.com/google/go-cmp@v0.3.1/cmp/options.go:133 +0x4d
github.com/google/go-cmp/cmp.Options.filter(0xc000e08480, 0x3, 0x4, 0xc00067ad80, 0x2ae8f00, 0x21ff240, 0x21ff240, 0xc000c3fc80, 0x194, 0x0, ...)
	github.com/google/go-cmp@v0.3.1/cmp/options.go:68 +0x27c
github.com/google/go-cmp/cmp.(*state).tryOptions(0xc00067ad80, 0x2ae8f00, 0x21ff240, 0x21ff240, 0xc000c3fc80, 0x194, 0x0, 0x0, 0x0, 0x3)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:268 +0xb5
github.com/google/go-cmp/cmp.(*state).compareAny(0xc00067ad80, 0x2a88240, 0xc000ee9630)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:224 +0x296
github.com/google/go-cmp/cmp.(*state).statelessCompare(0xc00067ad80, 0x2a88240, 0xc000ee9630, 0x1, 0xc00067ad80)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:203 +0x97
github.com/google/go-cmp/cmp.(*state).compareSlice(0xc00067ad80, 0x2ae8f00, 0x211f980, 0x211f980, 0xc000895620, 0x97, 0x211f980, 0xc000905680, 0x97)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:421 +0x2be
github.com/google/go-cmp/cmp.(*state).compareAny(0xc00067ad80, 0x2a88300, 0xc000eec780)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:254 +0xad5
github.com/google/go-cmp/cmp.(*state).compareInterface(0xc00067ad80, 0x2ae8f00, 0x21ff240, 0x21ff240, 0xc000c1ee90, 0x94, 0x21ff240, 0xc000c1eea0, 0x94)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:526 +0x47b
github.com/google/go-cmp/cmp.(*state).compareAny(0xc00067ad80, 0x2a88200, 0xc000ae2300)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:260 +0xdd2
github.com/google/go-cmp/cmp.(*state).compareMap(0xc00067ad80, 0x2ae8f00, 0x220b300, 0x220b300, 0xc000855ef0, 0x15, 0x220b300, 0xc000ab67e0, 0x15)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:500 +0x3db
github.com/google/go-cmp/cmp.(*state).compareAny(0xc00067ad80, 0x2a88300, 0xc000eec640)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:256 +0xe4d
github.com/google/go-cmp/cmp.(*state).compareInterface(0xc00067ad80, 0x2ae8f00, 0x21ff240, 0x21ff240, 0xc000c1edb0, 0x94, 0x21ff240, 0xc000c1edc0, 0x94)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:526 +0x47b
github.com/google/go-cmp/cmp.(*state).compareAny(0xc00067ad80, 0x2a88200, 0xc000a1b2c0)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:260 +0xdd2
github.com/google/go-cmp/cmp.(*state).compareMap(0xc00067ad80, 0x2ae8f00, 0x220b300, 0x220b300, 0xc000855d10, 0x15, 0x220b300, 0xc000ab6570, 0x15)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:500 +0x3db
github.com/google/go-cmp/cmp.(*state).compareAny(0xc00067ad80, 0x2a88300, 0xc000b60b40)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:256 +0xe4d
github.com/google/go-cmp/cmp.(*state).compareInterface(0xc00067ad80, 0x2ae8f00, 0x21ff240, 0x21ff240, 0xc000c31950, 0x94, 0x21ff240, 0xc000c31960, 0x94)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:526 +0x47b
github.com/google/go-cmp/cmp.(*state).compareAny(0xc00067ad80, 0x2a88200, 0xc000e4bb60)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:260 +0xdd2
github.com/google/go-cmp/cmp.(*state).compareMap(0xc00067ad80, 0x2ae8f00, 0x220b300, 0x220b300, 0xc0008362a0, 0x15, 0x220b300, 0xc0008362d0, 0x15)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:500 +0x3db
github.com/google/go-cmp/cmp.(*state).compareAny(0xc00067ad80, 0x2a6e780, 0xc000e084c0)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:256 +0xe4d
github.com/google/go-cmp/cmp.Equal(0x220b300, 0xc0008362a0, 0x220b300, 0xc0008362d0, 0xc000e59470, 0x3, 0x3, 0x0)
	github.com/google/go-cmp@v0.3.1/cmp/compare.go:107 +0x3a0
istio.io/operator/pkg/compare.YAMLCmpWithIgnore(0xc0006c0a80, 0x2a31, 0xc0006be000, 0x2a08, 0x0, 0x0, 0x0, 0xc000f03500, 0x28ab, 0x0, ...)
	istio.io/operator@v0.0.0-20191105031444-4d7f1ed76494/pkg/compare/compare.go:134 +0x4b7
istio.io/operator/cmd/mesh.checkUpgradeValues(0xc0006c0a80, 0x2a31, 0xc0006be000, 0x2a08, 0xc000f03500, 0x28ab, 0xc000e75d48)
	istio.io/operator@v0.0.0-20191105031444-4d7f1ed76494/cmd/mesh/upgrade.go:238 +0x81
istio.io/operator/cmd/mesh.upgrade(0xc0008a0acc, 0xc000aec7d0, 0xc000e75d48, 0x2a1a960, 0xc0000dc010)
	istio.io/operator@v0.0.0-20191105031444-4d7f1ed76494/cmd/mesh/upgrade.go:184 +0x896
istio.io/operator/cmd/mesh.UpgradeCmd.func1(0xc000768c80, 0xc000362cb0, 0x0, 0x7, 0x0, 0x0)
	istio.io/operator@v0.0.0-20191105031444-4d7f1ed76494/cmd/mesh/upgrade.go:105 +0x125
github.com/spf13/cobra.(*Command).execute(0xc000768c80, 0xc000362c40, 0x7, 0x7, 0xc000768c80, 0xc000362c40)
	github.com/spf13/cobra@v0.0.5/command.go:826 +0x460
github.com/spf13/cobra.(*Command).ExecuteC(0xc000acb680, 0x9, 0x9, 0xc000acb680)
	github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v0.0.5/command.go:864
main.main()
	istio.io/istio@/istioctl/cmd/istioctl/main.go:29 +0x76
```

--- a/pkg/name/name.go
+++ b/pkg/name/name.go
@@ -228,16 +228,21 @@ func GetFromTreePath(inputTree map[string]interface{}, path util.Path) (interfac
 		return val, true, nil
 	}
 	switch newRoot := val.(type) {
+	case string:
+		return nil, false, nil
 	case map[string]interface{}:
 		return GetFromTreePath(newRoot, path[1:])
 	case []interface{}:
 		for _, node := range newRoot {
-			nextVal, found, err := GetFromTreePath(node.(map[string]interface{}), path[1:])
-			if err != nil {
-				continue
-			}
-			if found {
-				return nextVal, true, nil
+			switch nt := node.(type) {
+			case map[string]interface{}:
+				nextVal, found, err := GetFromTreePath(nt, path[1:])
+				if err != nil {
+					continue
+				}
+				if found {
+					return nextVal, true, nil
+				}
 			}
 		}
 		return nil, false, nil

--- a/pkg/name/name.go
+++ b/pkg/name/name.go
@@ -19,6 +19,7 @@ import (
 	"reflect"
 
 	"istio.io/operator/pkg/apis/istio/v1alpha2"
+	"istio.io/operator/pkg/tpath"
 	"istio.io/operator/pkg/util"
 	"istio.io/pkg/log"
 )
@@ -220,34 +221,8 @@ func GetFromTreePath(inputTree map[string]interface{}, path util.Path) (interfac
 	if len(path) == 0 {
 		return nil, false, fmt.Errorf("path is empty")
 	}
-	val := inputTree[path[0]]
-	if val == nil {
-		return nil, false, nil
-	}
-	if len(path) == 1 {
-		return val, true, nil
-	}
-	switch newRoot := val.(type) {
-	case string:
-		return nil, false, nil
-	case map[string]interface{}:
-		return GetFromTreePath(newRoot, path[1:])
-	case []interface{}:
-		for _, node := range newRoot {
-			switch nt := node.(type) {
-			case map[string]interface{}:
-				nextVal, found, err := GetFromTreePath(nt, path[1:])
-				if err != nil {
-					continue
-				}
-				if found {
-					return nextVal, true, nil
-				}
-			}
-		}
-		return nil, false, nil
-	}
-	return GetFromTreePath(val.(map[string]interface{}), path[1:])
+	node, found := tpath.GetNodeByPath(inputTree, path)
+	return node, found, nil
 }
 
 // Namespace returns the namespace for the component. It follows these rules:

--- a/pkg/name/name_test.go
+++ b/pkg/name/name_test.go
@@ -111,6 +111,25 @@ n1
 			wantErr: false,
 		},
 		{
+			name: "path not in map array tree",
+			args: args{
+				inputTree: `
+a:
+  b:
+  - name: n1
+    value: v1
+  - name: n2
+    list:
+    - v21
+    - v22
+`,
+				path: util.Path{"a", "b", "unknown"},
+			},
+			want:    ``,
+			found:   false,
+			wantErr: false,
+		},
+		{
 			name: "node in map array tree",
 			args: args{
 				inputTree: `

--- a/pkg/name/name_test.go
+++ b/pkg/name/name_test.go
@@ -18,85 +18,149 @@ import (
 	"reflect"
 	"testing"
 
+	"github.com/ghodss/yaml"
+
 	"istio.io/operator/pkg/util"
 )
 
 func TestGetFromTreePath(t *testing.T) {
 	type args struct {
-		inputTree map[string]interface{}
+		inputTree string
 		path      util.Path
 	}
 
 	tests := []struct {
 		name    string
 		args    args
-		want    interface{}
+		want    string
 		found   bool
 		wantErr bool
 	}{
 		{
 			name: "found string node",
 			args: args{
-				inputTree: map[string]interface{}{
-					"k1": "v1",
-				},
+				inputTree: `
+k1: v1
+`,
 				path: util.Path{"k1"},
 			},
-			want:    "v1",
+			want: `
+v1
+`,
 			found:   true,
 			wantErr: false,
 		},
 		{
 			name: "found tree node",
 			args: args{
-				inputTree: map[string]interface{}{
-					"k1": map[string]interface{}{
-						"k2": "v2",
-					},
-				},
+				inputTree: `
+k1:
+ k2: v2
+`,
 				path: util.Path{"k1"},
 			},
-			want: map[string]interface{}{
-				"k2": "v2",
-			},
+			want: `
+k2: v2
+`,
 			found:   true,
 			wantErr: false,
 		},
 		{
 			name: "path is longer than tree depth, string node",
 			args: args{
-				inputTree: map[string]interface{}{
-					"k1": "v1",
-				},
+				inputTree: `
+k1: v1
+`,
 				path: util.Path{"k1", "k2"},
 			},
-			want:    nil,
+			want:    "",
 			found:   false,
 			wantErr: false,
 		},
 		{
 			name: "path is longer than tree depth, array node",
 			args: args{
-				inputTree: map[string]interface{}{
-					"k1": []interface{}{
-						"v1",
-					},
-				},
+				inputTree: `
+k1: v1
+`,
 				path: util.Path{"k1", "k2"},
 			},
-			want:    nil,
+			want:    "",
 			found:   false,
+			wantErr: false,
+		},
+		{
+			name: "string in map array tree",
+			args: args{
+				inputTree: `
+a:
+  b:
+  - name: n1
+    value: v1
+  - name: n2
+    list:
+    - v21
+    - v22
+`,
+				path: util.Path{"a", "b", "name"},
+			},
+			want: `
+n1
+`,
+			found:   true,
+			wantErr: false,
+		},
+		{
+			name: "node in map array tree",
+			args: args{
+				inputTree: `
+a:
+  b:
+    c: val1
+    list1:
+    - i1: val1
+    - i2: val2
+    - i3a: key1
+      i3b:
+        list2:
+        - i1: val1
+        - i2: val2
+        - i3a: key1
+          i3b:
+            i1: va11
+`,
+				path: util.Path{"a", "b", "list1", "i3b"},
+			},
+			want: `
+list2:
+- i1: val1
+- i2: val2
+- i3a: key1
+  i3b:
+    i1: va11
+`,
+			found:   true,
 			wantErr: false,
 		},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			got, found, err := GetFromTreePath(tt.args.inputTree, tt.args.path)
+			tree := make(map[string]interface{})
+			if err := yaml.Unmarshal([]byte(tt.args.inputTree), &tree); err != nil {
+				t.Fatal(err)
+			}
+			got, found, err := GetFromTreePath(tree, tt.args.path)
 			if (err != nil) != tt.wantErr {
 				t.Errorf("GetFromTreePath() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
-			if !reflect.DeepEqual(got, tt.want) {
+
+			var wantTree interface{}
+			if err := yaml.Unmarshal([]byte(tt.want), &wantTree); err != nil {
+				t.Fatal(err)
+			}
+
+			if !reflect.DeepEqual(got, wantTree) {
 				t.Errorf("GetFromTreePath() got = %v, want %v", got, tt.want)
 			}
 			if found != tt.found {

--- a/pkg/name/name_test.go
+++ b/pkg/name/name_test.go
@@ -1,0 +1,107 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package name
+
+import (
+	"reflect"
+	"testing"
+
+	"istio.io/operator/pkg/util"
+)
+
+func TestGetFromTreePath(t *testing.T) {
+	type args struct {
+		inputTree map[string]interface{}
+		path      util.Path
+	}
+
+	tests := []struct {
+		name    string
+		args    args
+		want    interface{}
+		found   bool
+		wantErr bool
+	}{
+		{
+			name: "found string node",
+			args: args{
+				inputTree: map[string]interface{}{
+					"k1": "v1",
+				},
+				path: util.Path{"k1"},
+			},
+			want:    "v1",
+			found:   true,
+			wantErr: false,
+		},
+		{
+			name: "found tree node",
+			args: args{
+				inputTree: map[string]interface{}{
+					"k1": map[string]interface{}{
+						"k2": "v2",
+					},
+				},
+				path: util.Path{"k1"},
+			},
+			want: map[string]interface{}{
+				"k2": "v2",
+			},
+			found:   true,
+			wantErr: false,
+		},
+		{
+			name: "path is longer than tree depth, string node",
+			args: args{
+				inputTree: map[string]interface{}{
+					"k1": "v1",
+				},
+				path: util.Path{"k1", "k2"},
+			},
+			want:    nil,
+			found:   false,
+			wantErr: false,
+		},
+		{
+			name: "path is longer than tree depth, array node",
+			args: args{
+				inputTree: map[string]interface{}{
+					"k1": []interface{}{
+						"v1",
+					},
+				},
+				path: util.Path{"k1", "k2"},
+			},
+			want:    nil,
+			found:   false,
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, found, err := GetFromTreePath(tt.args.inputTree, tt.args.path)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("GetFromTreePath() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("GetFromTreePath() got = %v, want %v", got, tt.want)
+			}
+			if found != tt.found {
+				t.Errorf("GetFromTreePath() found = %v, want %v", found, tt.found)
+			}
+		})
+	}
+}

--- a/pkg/tpath/tpath.go
+++ b/pkg/tpath/tpath.go
@@ -233,6 +233,35 @@ func IsLeafNode(treeNode interface{}) bool {
 	}
 }
 
+// GetNodeByPath returns the value at path from the given tree, or false if the path does not exist.
+func GetNodeByPath(treeNode interface{}, path util.Path) (interface{}, bool) {
+	if len(path) == 0 || treeNode == nil {
+		return nil, false
+	}
+	switch nt := treeNode.(type) {
+	case map[string]interface{}:
+		val := nt[path[0]]
+		if val == nil {
+			return nil, false
+		}
+		if len(path) == 1 {
+			return val, true
+		}
+		return GetNodeByPath(val, path[1:])
+	case []interface{}:
+		for _, nn := range nt {
+			np, found := GetNodeByPath(nn, path[1:])
+			if found {
+				return np, true
+			}
+		}
+		return nil, false
+	default:
+		return nil, false
+	}
+	panic("unreachable path")
+}
+
 // TODO Merge this into existing WritePathContext method (istio/istio#15494)
 // DeleteFromTree sets value at path of input untyped tree to nil
 func DeleteFromTree(valueTree map[string]interface{}, path util.Path, remainPath util.Path) (bool, error) {

--- a/pkg/tpath/tpath.go
+++ b/pkg/tpath/tpath.go
@@ -250,7 +250,7 @@ func GetNodeByPath(treeNode interface{}, path util.Path) (interface{}, bool) {
 		return GetNodeByPath(val, path[1:])
 	case []interface{}:
 		for _, nn := range nt {
-			np, found := GetNodeByPath(nn, path[1:])
+			np, found := GetNodeByPath(nn, path)
 			if found {
 				return np, true
 			}

--- a/pkg/tpath/tpath.go
+++ b/pkg/tpath/tpath.go
@@ -259,7 +259,6 @@ func GetNodeByPath(treeNode interface{}, path util.Path) (interface{}, bool) {
 	default:
 		return nil, false
 	}
-	panic("unreachable path")
 }
 
 // TODO Merge this into existing WritePathContext method (istio/istio#15494)

